exonum/src/runtime/rust/api.rs

```diff
@@ -160,10 +162,136 @@ impl ServiceApiScope {
+                    },
+                );
+                let result = handler(&state, query);
+                Box::new(result.into_future())
+            },
+        );
+        self
+    }
+
+    /// Same as `endpoint_mut`, but also add a warning about this endpoint being deprecated to the response.
+    pub fn deprecated_endpoint_mut<Q, I, F, R>(
+        &mut self,
+        name: &'static str,
+        discontinued_on: Option<Date<Utc>>,
+        handler: F,
+    ) -> &mut Self
+    where
+        Q: DeserializeOwned + 'static,
+        I: Serialize + 'static,
+        F: Fn(&ServiceApiState<'_>, Q) -> R + 'static + Clone + Send + Sync,
+        R: IntoFuture<Item = I, Error = crate::api::Error> + 'static,
+    {
+        let blockchain = self.blockchain.clone();
+        let descriptor = self.descriptor.clone();
+        self.inner.deprecated_endpoint_mut(
+            name,
+            discontinued_on,
+            move |query: Q| -> crate::api::FutureResult<I> {
+                let state = ServiceApiState::from_api_context(
+                    &blockchain,
+                    InstanceDescriptor {
+                        id: descriptor.0,
+                        name: descriptor.1.as_ref(),
+                    },
+                );
+                let result = handler(&state, query);
+                Box::new(result.into_future())
+            },
+        );
+        self
+    }
+
+    /// Creates an endpoint which will return "301 Moved Permanently" HTTP status code
+    /// to the incoming requests.
+    /// Response will include a "Location" header denoting a new location of the resource.
+    ///
+    /// `redirect_to` parameter must be a function that takes a request query as a parameter,
+    /// and returns a result of either `String` with the new location or `api::Error`.
+    ///
+    /// Example:
+    ///
+    /// ```rust,ignore
+    /// let public_scope = builder.public_scope();
+    /// // ...
+    /// let url_base = public_scope.url_base();
+    /// let moved_mutable_new_location = format!("{}/{}", &url_base, "new-endpoint-mut");
+    /// let moved_immutable_new_location = format!("{}/{}", &url_base, "new-endpoint");
+    /// public_scope
+    ///     .moved_permanently(
+    ///         "moved-mutable",
+    ///         move |_query: MutEndpointQuery| Ok(moved_mutable_new_location.clone()),
+    ///         EndpointMutability::Mutable,
+    ///     )
+    ///     .moved_permanently(
+    ///         "moved-immutable",
+    ///         move |query: EndpointQuery| {
+    ///             let query =
+    ///                 serde_urlencoded::to_string(query).expect("Unable to serialize query.");
+    ///     
+    ///             let new_location = format!("{}?{}", moved_immutable_new_location, query);
+    ///     
+    ///             Ok(new_location)
+    ///         },
+    ///         EndpointMutability::Immutable,
+    ///     );
+    /// ```
+    pub fn moved_permanently<Q, F>(
```

Comment:
Why are separate methods required once we have dedicated error types for `Moved Permanently` / `Gone` statuses? This looks unnecessarily complex and awkward to me.  (In fact, I was going to propose create these error types and handle endpoints via errors, but the prerequisites for this seem to be already here!)

Reply:
My original idea was to free users from creating any handlers by themselves for those cases, but seems like it turned out to be too confusing. I'll remove those methods and make returning error for those scenarios the intended use case.